tiny_dnn/node.h

```diff
@@ -197,40 +209,37 @@ inline std::shared_ptr<U> &operator<<(std::shared_ptr<T> &lhs,
   return rhs;
 }
 
-template <typename T, typename U>
-inline U &operator<<(const node_tuple<T> &lhs, U &rhs) {
-  for (serial_size_t i = 0; i < static_cast<serial_size_t>(lhs.nodes_.size());
-       i++) {
-    connect(&*lhs.nodes_[i], &*rhs, 0, i);
+template <typename T>
+inline T &operator<<(const layer_tuple<std::shared_ptr<layer>> &lhs, T &rhs) {
+  for (size_t i = 0; i < static_cast<size_t>(lhs.layers_.size()); i++) {
+    connect(&*lhs.layers_[i], &*rhs, 0, i);
   }
   return rhs;
 }
 
-template <typename T, typename U>
-inline node_tuple<T> &operator<<(U &lhs, const node_tuple<T> &rhs) {
-  for (serial_size_t i = 0; i < static_cast<serial_size_t>(rhs.nodes_.size());
-       i++) {
-    connect(&*lhs, &*rhs.nodes_[i], i, 0);
+template <typename T>
+inline layer_tuple<std::shared_ptr<layer>> &operator<<(
+  T &lhs, const layer_tuple<std::shared_ptr<layer>> &rhs) {
+  for (size_t i = 0; i < static_cast<size_t>(rhs.layers_.size()); i++) {
+    connect(&*lhs, &*rhs.layers_[i], i, 0);
   }
   return rhs;
 }
 
-template <typename T, typename U>
-inline U &operator<<(const node_tuple<T *> &lhs, U &rhs) {
-  for (serial_size_t i = 0; i < static_cast<serial_size_t>(lhs.nodes_.size());
-       i++) {
-    connect(lhs.nodes_[i], &rhs, 0, i);
+template <typename T>
+inline T &operator<<(const layer_tuple<layerptr_t> &lhs, T &rhs) {
+  for (size_t i = 0; i < static_cast<size_t>(lhs.layers_.size()); i++) {
+    connect(lhs.layers_[i], &rhs, 0, i);
   }
   return rhs;
 }
 
-template <typename T, typename U>
-inline node_tuple<T *> &operator<<(U &lhs, const node_tuple<T *> &rhs) {
-  for (serial_size_t i = 0; i < static_cast<serial_size_t>(rhs.nodes_.size());
-       i++) {
-    connect(&lhs, rhs.nodes_[i], i, 0);
+template <typename T>
+inline layer_tuple<layerptr_t> &operator<<(T &lhs,
+                                           const layer_tuple<layerptr_t> &rhs) {
+  for (size_t i = 0; i < static_cast<size_t>(rhs.layers_.size()); i++) {
+    connect(&lhs, rhs.layers_[i], i, 0);
   }
   return rhs;
```

Comment:
Seems like returning `rhs` as non-const reference is a problem since you pass it a const reference. Either add `const` qualifier to return type or drop it from parameter type (first is better, if possible)